File: oracle/js/src/priceSources/cmc.ts
// priceSources/cmc.ts — a minimal CoinMarketCap client, used only to price
// STEEM/USD and SBD/USD. Mirrors oracle/go/relayer/pricesource.go's
// CMCClient. Configured via ORACLE_CMC_API_KEY / ORACLE_CMC_BASE_URL.

import { toLegacyDecString } from "../decimalFmt";

export class CMCClient {
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(apiKey: string, baseUrl?: string) {
    this.apiKey = apiKey;
    this.baseUrl = (baseUrl || "https://pro-api.coinmarketcap.com").replace(/\/+$/, "");
  }

  /** Returns the latest USD price (as a LegacyDec.String()-formatted
   * string) for each of the given CoinMarketCap symbols, batched into a
   * single API call. Symbols CoinMarketCap doesn't return are simply absent
   * from the result. */
  async fetchUsdPrices(symbols: string[]): Promise<Record<string, string>> {
    if (symbols.length === 0) {
      return {};
    }
    const url = `${this.baseUrl}/v2/cryptocurrency/quotes/latest?symbol=${encodeURIComponent(symbols.join(","))}&convert=USD`;
    const resp = await fetch(url, {
      headers: { "X-CMC_PRO_API_KEY": this.apiKey, Accept: "application/json" },
    });
    if (!resp.ok) {
      throw new Error(`coinmarketcap: unexpected status ${resp.status}`);
    }

    // CoinMarketCap emits "price" as a raw JSON number literal. A plain
    // JSON.parse would round-trip it through float64 before we ever see it,
    // which can lose precision for prices with many significant digits —
    // the same failure mode toLegacyDecString's own doc comment warns
    // against. So the raw response text is patched to quote every "price"
    // number literal (preserving its exact source digits) BEFORE parsing;
    // json.Number in the Go client's encoding/json.Decoder achieves the
    // same "keep the original text" effect (see pricesource.go).
    const raw = await resp.text();
    const patched = raw.replace(
      /"price"\s*:\s*(-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?)/g,
      (_m, num: string) => `"price":${JSON.stringify(num)}`,
    );
    const parsed = JSON.parse(patched) as {
      data?: Record<string, Array<{ symbol: string; quote: Record<string, { price: string }> }>>;
    };

    const out: Record<string, string> = {};
    for (const sym of symbols) {
      const entries = parsed.data?.[sym];
      if (!entries || entries.length === 0) continue;
      const usd = entries[0].quote?.USD;
      if (!usd) continue;
      try {
        out[sym] = toLegacyDecString(usd.price);
      } catch {
        continue; // malformed price from upstream (e.g. exponent notation): skip rather than fail the batch
      }
    }
    return out;
  }
}

File: oracle/python/src/steemvm_oracle/pricesources/cmc.py
"""CoinMarketCap client, used only to price STEEM/USD and SBD/USD (Steem
has no native USD market). Ports ``oracle/go/relayer/pricesource.go``'s
``CMCClient``. Configured via ``ORACLE_CMC_API_KEY``/``ORACLE_CMC_BASE_URL``
-- see ``oracle/.env.example``.
"""

from __future__ import annotations

import json
from decimal import Decimal, InvalidOperation
from typing import Iterable

import httpx

DEFAULT_BASE_URL = "https://pro-api.coinmarketcap.com"


class CMCClient:
    def __init__(self, api_key: str, base_url: str = "", timeout: float = 15.0):
        self.api_key = api_key
        self.base_url = (base_url or DEFAULT_BASE_URL).rstrip("/")
        self._client = httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def fetch_usd_prices(self, symbols: Iterable[str]) -> dict[str, Decimal]:
        """Returns the latest USD price for each of the given CoinMarketCap
        symbols (e.g. "STEEM", "SBD"), batched into a single API call.
        Symbols CoinMarketCap doesn't return are simply absent from the
        result -- never raises for a missing/malformed individual symbol,
        only for a transport-level failure."""
        symbols = list(symbols)
        if not symbols:
            return {}

        url = f"{self.base_url}/v2/cryptocurrency/quotes/latest"
        params = {"symbol": ",".join(symbols), "convert": "USD"}
        headers = {"X-CMC_PRO_API_KEY": self.api_key, "Accept": "application/json"}
        resp = self._client.get(url, params=params, headers=headers)
        if resp.status_code != 200:
            raise RuntimeError(f"coinmarketcap: unexpected status {resp.status_code}")

        # parse_float=str preserves CoinMarketCap's exact textual price
        # representation, avoiding a float round-trip before it becomes a
        # Decimal -- mirrors the Go client's json.Number use (dec.UseNumber()).
        try:
            parsed = json.loads(resp.text, parse_float=str)
        except json.JSONDecodeError as exc:
            raise RuntimeError(f"coinmarketcap: invalid response: {exc}") from exc

        data = parsed.get("data") or {}
        out: dict[str, Decimal] = {}
        for sym in symbols:
            entries = data.get(sym)
            if not entries:
                continue
            usd = (entries[0].get("quote") or {}).get("USD") or {}
            price_raw = usd.get("price")
            if price_raw is None:
                continue
            try:
                out[sym] = Decimal(str(price_raw))
            except InvalidOperation:
                continue  # malformed price from upstream: skip rather than fail the batch
        return out
